### DISCUSSION TITLE
校验逻辑支持返回更多原始信息，拓展 IValidateResult 能力

### DISCUSSION THREAD

**Discussion by MeetzhDing** [alibaba/formily#3949] [Ideas]
## 现状
当前的校验逻辑，校验函数只能返回 IValidateResult 数据，数据中仅包含了一条 message 信息。对于一个复杂的校验逻辑，其中可能包含了许多其他的原始信息，这些原始信息可以用于其他的业务场景，甚至是复杂的表单展示。

典型的场景有：自定义滚动到指定页面位置（一个DomSelectorID）、自定义组件展示信息、原始的请求 Error。

https://github.com/alibaba/formily/blob/8861ef58fc098aef48309544a107ba2ee6a38876/packages/validator/src/types.ts#L17-L42

目前的类型设计上，message 部分情况为 string，部分情况为 any。any的情况如下。

https://github.com/alibaba/formily/blob/8861ef58fc098aef48309544a107ba2ee6a38876/packages/core/src/types.ts#L121-L143

这里虽然可以通过类型 hack，通过 message 返回一个 object 类型，但是这样会导致目前已有的 UI 适配库都需要改动，而且类型也不严格准确。

## 提议
在 IValidateResult 中，新增 info 字段，允许返回 IValidateResult 时额外捎带一个信息字段，可以在字段中包含更多信息。

## 主要改动点
改动 /packages/validator/src/validator.ts 文件，同时处理 message 和 info 信息。info 字段类型为 any。
https://github.com/alibaba/formily/blob/8861ef58fc098aef48309544a107ba2ee6a38876/packages/validator/src/validator.ts#L18-L39

**Comment by MeetzhDing** [comment]
@janryWang 这个问题可以帮忙看看吗？一些复杂的表单场景，会有一些特殊的错误处理逻辑。

这个改动个人评估是兼容性变更，且开发工作量不大。如果觉得建议可行，愿意采纳，我也可以进行开发与PR.

**Comment by janryWang** [comment]
先聊聊你们的需求场景？具体是要做什么？为什么需要附带信息？因为我理解校验信息的唯一消费方就是组件

**Reply by MeetzhDing** [reply]
典型的场景有：
- 自定义滚动到指定页面位置（一个DomSelectorID）：校验不通过的原因可能是复杂的，用户看到错误文案后，可能需要点击跳转到某个具体的区域
- 自定义组件展示信息：校验信息本身可能是多元的，不一定非要是个纯字符串文案，可能是两句文案，errorMsg + errorDetail，甚至包含了 error 的信息，包含了错误的解决方案信息，用更具丰富的UI展示出来
- 校验出错本身是会有侧边栏组件区消费，侧边栏是整个表单的目录。这个时候错误就不止有错误文案，还有错误的位置，错误的业务类型，等等情况。

我后续仔细看了这块的类型代码，个人觉得比较简单的方法，就是   让 Schema 库中的 Message 类型，直接和 Core 中的类型保持一个，是一个 any 类型。
这样组件兼容、复杂场景都能支持，且改动量也最小。

@janryWang

**Comment by MeetzhDing** [comment]
Schema 和 Core 的类型这块，其实还有一个问题：

现在很多地方是直接包了一层 stringify 的， 就是有个自带的  `| ({} & string)`
这个东西的功能我是理解的，但是对于要进行源码开发的情况，这个东西让整个类型系统弱化了非常多。

比如 setDisplay  setPattern，其实现在就什么参数都是合法的了，但是本身TypeScript应该提供这种校验能力的。
这种逻辑对于一些使用 schema/tsx 源码开发的情况，影响就很大了。
原生的类型可能必须要被重写才能够作为一个较为可靠的类型定义来使用，是否可以考虑：

类型本身是不包含 `| ({} & string)` 的，对外导出的类型，根据路径分为两套，一个是有的，一个是没有的？
就比如代码为：
```ts
export ISchema = {}
export ISchemaStringify = Stringify<ISchema>
```

**Comment by daisymark** [comment]
请问这个有进展吗? 我也需要这个能力。这个大大增加了可扩展能力。

**Reply by MeetzhDing** [reply]
@daisymark 
官方没有处理。
因为我们这边的 UI 是自己的组件库，后来直接把 message 返回一个 object 来处理了。

Formily 整体和字段相关的类型都很弱，我们自己用工具函数包了一层，把类型转掉了。

**Reply by daisymark** [reply]
麻烦有参考代码吗我借鉴下

**Reply by daisymark** [reply]
定制x-rules的话，在render时候是不是也是需要整个UI需要自己实现了~

**Reply by daisymark** [reply]
![image](https://github.com/user-attachments/assets/df4cde35-08b7-44b4-b893-36e5d8544a62)
我看官网文档支持JSX

**Reply by MeetzhDing** [reply]
只需要修改 FormItem 组件就可以，message 类型为 {text: string; other: '...'}

FormItem 原本读取 field.message 就是一个对象，会有问题
https://github.com/alibaba/formily/blob/formily_next/packages/antd/src/form-item/index.tsx#L424

修改 takeMassage 逻辑，对massage做额外的解析处理即可

**Reply by daisymark** [reply]
好的，多谢！
